Set Stroke Property and a few optimizations

rektuiemforameme opened the pull request:
-Added the Set Stroke Property operator to GP Extras. It functions much like the stroke hardness modifier, except that it sets the stroke hardness, stroke line_width,  point pressure,  or point strength to a given value instead of controlling it incrementally with the mouse wheel. Useful if you already know what value you want to set so you don't have to mess with the wheel. It has three rows, with the top and bottom rows having set values of 1 and 0, and the middle row having a field to input the value to set. At first I just had one row with an input field, but after a while of using it, I found that I would often go between 0 and 1 so decided to add the set values for convenience so you don't have to keep setting the field.

-Changed init_selected to get_selected_points, which returns the list of points instead of setting a global. 

-Removed global variable selected_points in favor of variables local to each operator so that the variable doesn't stick around taking up memory when it's not being used. Variables don't usually take up a lot of memory, but this is a list that could store 1000+ items, and it all adds up. Also, globals are honestly just bad in general and should only be used if needed imo. 

-Changed get_selected_points to use use_multiedit so you can use the operators on multiple frames at once if Multiframe is enabled. 

-Changed get_selected_points to use list comprehension instead of clearing and appending points. The new method is faster, taking 73% of the time of appending in my albeit limited test. Small optimization because of how fast it is, but everything adds up. 

-Adjusted hardness operator's description to be more accurate. It had strength and pressure mixed up. 

-Cleaned up panel ui code a little. You don't need new columns for each button, you can just put them directly in the row, box or layout. 

-Made operators in panel ui centered horizontally. This is just an aesthetic choice I guess.

rektuiemforameme committed:
Set Stroke Property and a few optimizations

-Added the Set Stroke Property operator to GP Extras. It functions much like the stroke hardness modifier, except that it sets the stroke hardness, stroke line_width,  point pressure,  or point strength to a given value instead of controlling it incrementally with the mouse wheel. Useful if you already know what value you want to set so you don't have to mess with the wheel. It has three rows, with the top and bottom rows having set values of 1 and 0, and the middle row having a field to input the value to set. At first I just had one row with an input field, but after a while of using it, I found that I would often go between 0 and 1 so decided to add the set values for convenience so you don't have to keep setting the field.
-Changed init_selected to get_selected_points, which returns the list of points instead of setting a global.
-Removed global variable selected_points in favor of variables local to each operator so that the variable doesn't stick around taking up memory when it's not being used. Variables don't usually take up a lot of memory, but this is a list that could store 1000+ items, and it all adds up. Also, globals are honestly just bad in general and should only be used if needed imo.
-Changed get_selected_points to use use_multiedit so you can use the operators on multiple frames at once if Multiframe is enabled.
-Changed get_selected_points to use list comprehension instead of clearing and appending points. The new method is faster, taking 73% of the time of appending in my albeit limited test. Small optimization because of how fast it is, but everything adds up.
-Adjusted hardness operator's description to be more accurate. It had strength and pressure mixed up.
-Cleaned up panel ui code a little. You don't need new columns for each button, you can just put them directly in the row, box or layout.
-Made operators in panel ui centered horizontally. This is just an aesthetic choice I guess.

PongBuster commented:
Good idea with the optimization and set properties window. I have been
trying out your changes and have some thoughts.

Unfortunately Blender has not implemented the hardness value on a linear
scale, 0.5 is not 50%. It appears to be more like 5 to 10% ( log scale
maybe). That is why I implemented it with middle mouse wheel so I could
'home' in on the look I wanted. The radius and transparency was an after
thought.

They dovetailed though when you account that the stoke points may have
thickness and pressure values set. I.e a line fading/tapering out. A single
setting would wipe out those values (like normalize line thickness/opacity
commands in stroke menu). However, maybe the best of both worlds can be
implemented.

Scroll middle mouse button would add/subtract values and changing the
values in the property window would set them uniform. Not sure if this can
be done in that property window. Could be in a custom dialog though.

In the property/dialog the hardness value should have a max min range 0 to
1. This should effect the hardness on a linear scale. Same with
pressure/transparency. Line width/radius is messed up right now as it also
changes hardness. It should limit to 0 min and 300 max.

From a ui design point, I would limit your sidebar to 1 button (the quick
set buttons are over kill and it looks copy cluttered). Don't wipe out my
button just yet. :)

My suggestions going forward:
- add a stroke menu command Normalize hardness that sets hardness of all
selected stokes to 1.
- when clicking on your sidebar panel button a  property/dialog window is
displayed initialized with the selected stroke's hardness, radius/line
width, pressure/transparency values
- a modal callback is also run that adds/subtracts like mine
- if a value is overriden in the property/dialog it sets all
selected/strokes to that changed value.

Does this sound reasonable to you?

On Mon., Nov. 13, 2023, 1:53 a.m. Matt Thompson, ***@***.***>
wrote:

> -Added the Set Stroke Property operator to GP Extras. It functions much
> like the stroke hardness modifier, except that it sets the stroke hardness,
> stroke line_width, point pressure, or point strength to a given value
> instead of controlling it incrementally with the mouse wheel. Useful if you
> already know what value you want to set so you don't have to mess with the
> wheel. It has three rows, with the top and bottom rows having set values of
> 1 and 0, and the middle row having a field to input the value to set. At
> first I just had one row with an input field, but after a while of using
> it, I found that I would often go between 0 and 1 so decided to add the set
> values for convenience so you don't have to keep setting the field.
>
> -Changed init_selected to get_selected_points, which returns the list of
> points instead of setting a global.
>
> -Removed global variable selected_points in favor of variables local to
> each operator so that the variable doesn't stick around taking up memory
> when it's not being used. Variables don't usually take up a lot of memory,
> but this is a list that could store 1000+ items, and it all adds up. Also,
> globals are honestly just bad in general and should only be used if needed
> imo.
>
> -Changed get_selected_points to use use_multiedit so you can use the
> operators on multiple frames at once if Multiframe is enabled.
>
> -Changed get_selected_points to use list comprehension instead of clearing
> and appending points. The new method is faster, taking 73% of the time of
> appending in my albeit limited test. Small optimization because of how fast
> it is, but everything adds up.
>
> -Adjusted hardness operator's description to be more accurate. It had
> strength and pressure mixed up.
>
> -Cleaned up panel ui code a little. You don't need new columns for each
> button, you can just put them directly in the row, box or layout.
>
> -Made operators in panel ui centered horizontally. This is just an
> aesthetic choice I guess.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/PongBuster/Blender-Addons/pull/1
> Commit Summary
>
>    - d8c5129
>    <https://github.com/PongBuster/Blender-Addons/pull/1/commits/d8c5129bd1e876fa744843d68665b6787e684e80>
>    Set Stroke Property and a few optimizations
>
> File Changes
>
> (1 file <https://github.com/PongBuster/Blender-Addons/pull/1/files>)
>
>    - *M* gp_GreasePencilExtras.py
>    <https://github.com/PongBuster/Blender-Addons/pull/1/files#diff-1b497c630280ae340c9366429d847febac15406d73c6535fc33f79d20f249d12>
>    (238)
>
> Patch Links:
>
>    - https://github.com/PongBuster/Blender-Addons/pull/1.patch
>    - https://github.com/PongBuster/Blender-Addons/pull/1.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PongBuster/Blender-Addons/pull/1>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A56G5LPZJDCZIWV2XDHEY4LYEGYVVAVCNFSM6AAAAAA7IT3XV2VHI2DSMVQWIX3LMV43ASLTON2WKOZRHE4DSOJTHA4TEMI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

PongBuster reviewed (changes requested):
- add a stroke menu command Normalize hardness that sets hardness of all
selected stokes to 1.
- when clicking on your sidebar panel button a  property/dialog window is
displayed initialized with the selected stroke's hardness, radius/line
width, pressure/transparency values
- a modal callback is also run that scales the values rather with the middle mouse wheel than setting them outright
- if a value is overriden in the property/dialog it sets all
selected/strokes to that changed value.

rektuiemforameme commented:
> That is why I implemented it with middle mouse wheel so I could
'home' in on the look I wanted

I agree. I'm not trying to make the mouse wheel operator obsolete; I think both ways of setting the various properties are better in different situations. 
> In the property/dialog the hardness value should have a max min range 0 to 1

The property doesn't have a maximum because I wanted to use one field for all settings, and both line width and pressure can go over 1. I removed line width, but I like being able to set the pressure higher than 1 because it's easier to make a small number of points thicker than to make all of the rest thinner and then increase the line width.
> This should effect the hardness on a linear scale

I checked [blender's code](https://github.com/blender/blender/blob/main/source/blender/draw/intern/shaders/common_gpencil_lib.glsl#L38) and changed the hardness set code to optionally use ```pow(self.value, 0.1)``` which produces this result which I feel is much more linear:
![linear hardness](https://i.imgur.com/ouoxPdg.png)
The main downside with this is that our hardness values won't be compatible with the brush's hardness settings since they're on different scales. For example, in one of my current projects I am using a hardness of 0.8 as the default for all lines. If I animate the hardness and then want to set it back to standard, 0.8 won't be the value I need anymore. I'll have to do some math to figure out what the linear value would be to get 0.8 for the stroke's actual hardness. As such, I feel like it should be optional. I added an option for the operator called Linear Scale that is off by default. You can enable it in the redo panel, or by shift-clicking the hardness set button. The option is only available in redo if the mode is Hardness, and saves its state so that it is the same the next time you call the operator. Not sure this is the best way to do it. Let me know if you can think of a better alternative.

As a side note, the way blender draws hardness is fucked. Honestly, whoever coded this shit must have been drunk.
> add a stroke menu command Normalize hardness that sets hardness of all selected stokes to 1.

I added an option to the stroke menu "Set Hardness" that calls set_stroke_property with hardness and a default of 1. It does the same thing as what you asked for, but I feel like Set is a more accurate name for what it does than Normalize, especially since you can change its value in the redo panel.
> when clicking on your sidebar panel button a  property/dialog window is displayed initialized with the selected stroke's hardness, radius/line width, pressure/transparency values

I'm not sure how this would work when you have multiple strokes or points selected that might have different property values. Would it just be initialized to the most recently selected stroke and point? I'm not sure that's even possible, since there's no such thing as the 'active' stroke or point.
The operator opens a redo panel that you can use to change the value and it sets the properties when you change the value like you want, although it's just one value and is initialized to the value set in the panel, not the stroke's current property. I feel like that's preferable though. I doubt I'll be wanting to set multiple properties at once, and if I do I can just do the operator twice. 
![Op Redo Panel](https://i.imgur.com/RE689rb.png)
> a modal callback is also run that scales the values rather with the middle mouse wheel than setting them outright

Are you suggesting we basically combine the mouse wheel operator with the button quick set? Like I said at the beginning, I think they have different use cases. The whole point of setting to a property for me was to make it as quick as possible, with as few inputs required as possible. Preferably, just input a value and click. Honestly, it would just slow me down if I then have to click out of a modal or go to a secondary window. 

I removed the 0 and 1 set rows and the line width option, so now it's just the field and three buttons. You can set to 1 or 0 by ctrl or alt clicking the buttons, which I feel is better anyway. I think having multiple buttons saves time, and with just the one row I don't think it looks cluttered anymore.
![Changed Panel](https://i.imgur.com/kvP1Ily.png)

rektuiemforameme committed:
Minor Adjustments to stroke property set

Added Linear Scale option to hardness set.
Ctrl or Alt clicking a stroke prop set button will use 1 or 0 as the set value.
Removed 0 and 1 stroke prop set rows and stroke set line width from panel.
Added Set Hardness to stroke edit menu.

PongBuster commented:
Strokes have a selected_index value which contains the order the strokes were picked. The rule could be if only one stroke is picked then initialize, but even if more then one is selected they are all going to be wiped out anyways with the same setting so initializing to the first selected one is better then 0 or 1 visually.

PongBuster commented:
My reasoning is this: unless setting to 0 or 1, the desired hardness value is subjective visually (especially because of the wacky scaling value). I would use the sliding mechanism on the value property to scale the hardness to the look I want. i.e. 0.27.

Not sure if this is possible but imagine the float numeric property is disabled normally, the user would click the set hardness button, the property is initialized from the selected stroke(s) hardness value and enabled. They could then slide the value to desired amount and release or type in a value to set the hardness.  When the float property box loses focus it becomes disabled. The hardness button would be outlined/highlighted indicating that it was the value that was being modified.

BTW, clicking the Pressure button right now makes the selected stroke disappear. I need to ALT-S to scale the stroke thickness back up. I am thinking maybe they should be labelled Transparency and Thickness. Pressure and Strength is a bit confusing as it could mean either or.

rektuiemforameme commented:
> I would use the sliding mechanism on the value property to scale the hardness to the look I want. i.e. 0.27.

But isn't that what the scrollwheel hardness op is for? You can also do this from the redo panel. Altogether, your proposal sounds useful, but it's not what I want this tool to be. Like I said, I use 0.8 as a standard hardness value, so it's more useful to me to keep 0.8 in that field than to have it update to the stroke's current hardness and then I have to type in 0.8. Maybe your proposal should be a different tool with a button next to the hardness operator that opens a popup?

> BTW, clicking the Pressure button right now makes the selected stroke disappear. I need to ALT-S to scale the stroke thickness back up.

Clicking the Pressure button is working fine for me. Are you sure you're not Alt clicking the button, which would set it to 0 instead of the property value?
![Pressure press](https://i.imgur.com/tFNfN0f.png)
> I am thinking maybe they should be labelled Transparency and Thickness. Pressure and Strength is a bit confusing as it could mean either or.

I adjusted the descriptions to use "Pressure(Thickness)" and "Strength(Transparency)". 

I'm opening a new PR with this and a fix to the linear scaling. If you're still having issues with the pressure, let me know there.

Future matt here. I thought I posted this comment hours ago.
